// edge-tts-node/dist/Headers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VOICE_HEADERS = exports.WSS_HEADERS = exports.BASE_HEADERS = exports.CHROMIUM_MAJOR_VERSION = void 0;
/// headers
const CHROMIUM_MAJOR_VERSION = "130";
exports.CHROMIUM_MAJOR_VERSION = CHROMIUM_MAJOR_VERSION;
const BASE_HEADERS = {
    "User-Agent": `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/${CHROMIUM_MAJOR_VERSION}.0.0.0 Safari/537.36 Edg/${CHROMIUM_MAJOR_VERSION}.0.0.0`,
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-US,en;q=0.9",
};
exports.BASE_HEADERS = BASE_HEADERS;
const WSS_HEADERS = {
    Pragma: "no-cache",
    "Cache-Control": "no-cache",
    Origin: "chrome-extension://jdiccldimpdaibmpdkjnbmckianbfold",
    ...BASE_HEADERS,
};
exports.WSS_HEADERS = WSS_HEADERS;
const VOICE_HEADERS = {
    Authority: "speech.platform.bing.com",
    "Sec-CH-UA": `" Not;A Brand";v="99", "Microsoft Edge";v="${CHROMIUM_MAJOR_VERSION}", "Chromium";v="${CHROMIUM_MAJOR_VERSION}"`,
    "Sec-CH-UA-Mobile": "?0",
    Accept: "*/*",
    "Sec-Fetch-Site": "none",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Dest": "empty",
    ...BASE_HEADERS,
};
exports.VOICE_HEADERS = VOICE_HEADERS;
